query.go

```diff
@@ -82,174 +161,212 @@ func (dht *IpfsDHT) runDisjointQueries(ctx context.Context, d int, target string
 
 	// distribute the shuffled K closest peers as seeds among the "d" disjoint queries
 	for i := 0; i < len(seedPeers); i++ {
-		queries[i%d].localPeers.Add(seedPeers[i])
+		queries[i%d].seedPeers = append(queries[i%d].seedPeers, seedPeers[i])
 	}
 
 	// start the "d"  disjoint queries
+	queryDone := make(chan struct{}, d)
 	for i := 0; i < d; i++ {
 		query := queries[i]
 		go func() {
-			strictParallelismQuery(query)
+			query.runWithGreedyParallelism()
 			queryDone <- struct{}{}
 		}()
 	}
 
-loop:
 	// wait for all the "d" disjoint queries to complete before we return
+	// XXX: Waiting until all queries are done is a vector for DoS attacks:
+	// The disjoint lookup paths that are taken over by adversarial peers
+	// can easily be fooled to go on forever.
+	numQueriesComplete := 0
 	for {
-		select {
-		case <-queryDone:
-			numQueriesComplete++
-			if numQueriesComplete == d {
-				break loop
-			}
-		case <-ctx.Done():
-			break loop
+		<-queryDone
+		numQueriesComplete++
+		if numQueriesComplete == d {
+			break
 		}
 	}
 
-	return queries, nil
+	res := dht.constructLookupResult(queries, targetKadID)
+	return res, nil
 }
 
-// TODO This function should be owned by the DHT as it dosen't really belong to "a query".
-// scorePeerByDistanceAndLatency scores a peer using metrics such as connectendness of the peer, it's distance from the key
-// and it's current known latency.
-func (q query) scorePeerByDistanceAndLatency(p peer.ID, distanceFromKey *big.Int) interface{} {
-	connectedness := q.dht.host.Network().Connectedness(p)
-	latency := q.dht.host.Peerstore().LatencyEWMA(p)
-
-	var c int64
-	switch connectedness {
-	case network.Connected:
-		c = 1
-	case network.CanConnect:
-		c = 5
-	case network.CannotConnect:
-		c = 10000
-	default:
-		c = 20
+// constructLookupResult takes the query information and uses it to construct the lookup result
+func (dht *IpfsDHT) constructLookupResult(queries []*query, target kb.ID) *lookupResult {
```

Comment:
@petar added this function to construct the result object we're going to return. As you can see the disjoint paths make the code a little more complex. If it's too much of a pain to maintain it's likely we'll just remove it.